File: Personal/downloadstudy.py
import requests, os
import io
import time
from tqdm import tqdm
from dotenv import load_dotenv

def get_study_ids(orthanc_url, basic_auth, study_instance_uid):
    start_time = time.time()

    # Construct the URL for the query
    query_url = f"{orthanc_url}/tools/find"

    # Construct the query parameters
    query_params = {
        "Level": "Study",
        "Query": {
            "StudyInstanceUID": study_instance_uid
        }
    }

    # Measure time taken to establish connection and send request
    connection_start = time.time()
    response = requests.post(query_url, json=query_params, auth=("admin", "password"))
    connection_time = time.time() - connection_start
    print(f"Time taken to establish connection and send request: {connection_time:.2f} seconds")

    # Check if request was successful
    if response.status_code == 200:
        # Parse the JSON response
        parsing_start = time.time()
        studies = response.json()
        parsing_time = time.time() - parsing_start
        print(f"Time taken to parse response: {parsing_time:.2f} seconds")
        
        elapsed_time = time.time() - start_time
        print(f"Total time taken to retrieve Study ID: {elapsed_time:.2f} seconds")
        
        if studies:
            # Return the Study ID
            return studies[0]
        else:
            return None
    else:
        # Raise an exception if the request failed
        response.raise_for_status()

def download_study_as_zip(orthanc_url, basic_auth, study_id):
    start_time = time.time()
    
    # Construct the URL to download the study archive
    archive_url = f"{orthanc_url}/studies/{study_id}/media"

    # Measure time to start connection for downloading the archive
    connection_start = time.time()
    with requests.get(archive_url, auth=("admin", "password"), stream=True) as response:
        connection_time = time.time() - connection_start
        print(f"Time taken to establish connection and start download: {connection_time:.2f} seconds")
        
        response.raise_for_status()
        
        # Total size in bytes
        total_size = int(response.headers.get('content-length', 0))
        block_size = 1024  # 1 KiB blocks

        # Initialize the tqdm progress bar with percentage
        tqdm_bar = tqdm(total=total_size, unit='B', unit_scale=True, desc='Downloading', ncols=100, ascii=True, leave=True, bar_format='{l_bar}{bar} | {percentage:3.0f}%')

        # Measure time to download and write content in chunks
        download_start = time.time()
        zip_file_content = io.BytesIO()

        for data in response.iter_content(block_size):
            tqdm_bar.update(len(data))
            zip_file_content.write(data)
        
        download_time = time.time() - download_start
        tqdm_bar.close()
        print(f"Time taken to download and write content: {download_time:.2f} seconds")
    
    # Calculate time taken
    elapsed_time = time.time() - start_time
    print(f"Total time taken to download study as ZIP: {elapsed_time:.2f} seconds")

    # Return the content of the ZIP file
    return zip_file_content.getvalue()

def download_study(orthanc_url, basic_auth, study_instance_uid):
    start_time = time.time()
    
    # Get the Study ID using the provided Study Instance UID
    study_id = get_study_ids(orthanc_url, basic_auth, study_instance_uid)

    if not study_id:
        raise ValueError("StudyInstanceUID not found")

    # Download the study as a ZIP file
    zip_file_content = download_study_as_zip(orthanc_url, basic_auth, study_id)

    # Measure time to create a BytesIO object from the zip file content
    bytesio_start = time.time()
    zip_file = io.BytesIO(zip_file_content)
    bytesio_time = time.time() - bytesio_start
    print(f"Time taken to create BytesIO object: {bytesio_time:.2f} seconds")

    # Calculate total time taken
    total_time = time.time() - start_time
    print(f"Total time taken for the whole process: {total_time:.2f} seconds")

    # Return the BytesIO object and the study ID (for naming the file)
    return zip_file, study_id
load_dotenv()

orthanc_url = os.getenv('ORTHANC_LINK')
basic_auth = os.getenv('BASIC_AUTH')
study_instance_uid = os.getenv('STUDY_INSTANCE_UID')

# Test the function
if __name__ == "__main__":
    try:
        # Call the function
        zip_file, study_id = download_study(orthanc_url, basic_auth, study_instance_uid)

        # Save the file to disk (for testing purposes)
        with open(f'{study_id}.zip', 'wb') as f:
            f.write(zip_file.getvalue())
        print(f"Study {study_id} downloaded successfully as {study_id}.zip")

    except Exception as e:
        print(f"Error: {e}")

File: Personal/downloadstudy.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strings"
)

func getStudyIDs(orthancURL, basicAuth, studyInstanceUID string) ([]string, error) {
	// Construct the URL for the query
	queryURL := fmt.Sprintf("%s/tools/find", orthancURL)

	// Construct the query parameters
	queryParams := map[string]interface{}{
		"Level": "Study",
		"Query": map[string]string{
			"StudyInstanceUID": studyInstanceUID,
		},
	}

	// Convert queryParams to JSON
	queryParamsJSON, err := json.Marshal(queryParams)
	if err != nil {
		return nil, err
	}

	// Construct the request
	req, err := http.NewRequest("POST", queryURL, strings.NewReader(string(queryParamsJSON)))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", basicAuth)

	// Send the request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Read the response body
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// Print raw response body for debugging
	fmt.Printf("Raw response body: %s\n", string(body))

	// Check if request was successful
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to retrieve study IDs: %s", resp.Status)
	}

	// Parse the JSON response as an array of strings
	var studyIDs []string
	if err := json.Unmarshal(body, &studyIDs); err != nil {
		return nil, err
	}

	if len(studyIDs) == 0 {
		return nil, fmt.Errorf("no studies found")
	}

	return studyIDs, nil
}

func downloadStudyAsZip(orthancURL, basicAuth, studyID string) ([]byte, error) {
	archiveURL := fmt.Sprintf("%s/studies/%s/archive", orthancURL, studyID)

	// Construct the request
	req, err := http.NewRequest("GET", archiveURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", basicAuth)

	// Send the request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Check if request was successful
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to download study: %s", resp.Status)
	}

	// Read the response body
	return ioutil.ReadAll(resp.Body)
}

func main() {
	// Example usage
	orthancURL := "http://localhost:8042"
	basicAuth := "Basic YWRtaW46cGFzc3dvcmQ=" // Replace with your Base64 encoded credentials
	studyInstanceUID := "1.3.12.2.1107.5.1.7.107889.30000024081417115758500000006"

	studyIDs, err := getStudyIDs(orthancURL, basicAuth, studyInstanceUID)
	if err != nil {
		log.Fatalf("Error getting study IDs: %v", err)
	}

	fmt.Printf("Study IDs: %v\n", studyIDs)

	if len(studyIDs) > 0 {
		studyID := studyIDs[0]
		zipData, err := downloadStudyAsZip(orthancURL, basicAuth, studyID)
		if err != nil {
			log.Fatalf("Error downloading study: %v", err)
		}

		// Save zip file to disk
		err = ioutil.WriteFile("study.zip", zipData, 0644)
		if err != nil {
			log.Fatalf("Error saving zip file: %v", err)
		}
	}
}
